# SCIKIT LEARN WORKFLOW
## END TO END SCIKIT LEARN WORKFLOW

What is end to end means? 

One thing to note though Scikit Learn is such big library thus only call what you need will be a better approach to practical use.

### Random Forest Classifier Workflow for Classifying Heart Disease.
#### 1. Get the data ready
I will use the heart disease data. Remember that it was stored in csv format. First I need to import it first. Pandas will be the appropriate library to get and process the data to form Data Frame.

In [1]:
# get the data heart-disease.csv
import pandas as pd
heart_disease_data = pd.read_csv('../data/heart-disease.csv')
heart_disease_data

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

***NOTE***
The heart_disease_data consist of all patient characteristics except for the target column. Target column is the end result whether the patient has heart_disease or not and we want to seek the connection with each character.

In [2]:
# first we need to separate the target column from other data
x_data = heart_disease_data.drop('target', axis=1)
x_data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  
0   0     1  
1   0     2  
2   0     2  
3   0     2  
4   0     2

In [3]:
# now the target label will be in the y_data
y_data = heart_disease_data['target']
y_data.head()

0    1
1    1
2    1
3    1
4    1
Name: target, dtype: int64

***NOTE:***
now I need to split the data frame in the x_data and y_data to do this I will use the scikit learn libs

The way to import is a bit different since it is named sklearn thus to import I need import sklearn.

Now as this library is quite bid I need to be specific. This time it is model_selection to be exact I want to randomly split the data frame heart_disease_data y_data and x_data. I will use the train_test_split library

In [4]:
# import the library
# https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
from sklearn.model_selection import train_test_split

# just to be safe if we need to handle some matrices I will import numpy as well
import numpy as np

In [5]:
x_train, x_test, y_train, y_test = train_test_split(x_data, y_data)
# let's test the result but since too big I will just validate each size
x_train.shape, y_train.shape, x_test.shape, y_test.shape

((227, 13), (227,), (76, 13), (76,))

***NOTE:***
The data (shape) above is important. 

I need to validatae that the x_train and y_train first size is the same (in this case 227).

The same case is true for the x_test and y_test (in this case 76)

It means will be 227 data for train and 76 for testing that chosen randomly.

#### 2. Choose the model and hyperparameters
Now this is the very heart of the Scikit Learn. A variety of models or in machine learning is known as classifiers. However, in Scikit documentation it is more commonly known as estimators. 

TODO So what is hyperparameters??

Next we will use Random Forest Classifier: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html?highlight=forest#sklearn.ensemble.RandomForestClassifier

 learn about Random Forest Classifier model: https://en.wikipedia.org/wiki/Random_forest
 It is said that this is part of the ensemble learning method thus it is put in the Sklearn ensemble libs

In [6]:
# import the Random Forest 
from sklearn.ensemble import RandomForestClassifier
# too long the name thus better put it into a variable
clf = RandomForestClassifier()

In [7]:
clf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

#### 3. Fit the model to the data and use it to make predictions
Fitting the model on the data involves passing it the data and asking it to figure out the patterns.

If there are labels (supervised learning), the model tries to work out the relationship between the data and the labels.

If there are no labels (unsupervised learning), the model tries to find patterns and group similar samples together.

First I need to build the forest of trees model from the training set, in this case is **x_train and y_train.**

Read more about the forest of the random forest classifier's fit method: https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.fit



In [8]:
# making random forest classifier 
clf.fit(x_train, y_train)

RandomForestClassifier()

***NOTE:***
This clf.fit does not need to be stored in variable. I think it will just use the clf object to store the data from both side of randomly selected train data. 


Here clf = RandomForestClassifier is an **object**. It is just like the **DataFrame** in Pandas. You can modify directly the instatiation of this DataFrame (in this case heart_disease_data) which you can cut, process, etc. 

#### Use the model to make a prediction
The whole point of training a machine learning model is to use it to make some kine of prediction in the future. 

Once our model instance is trained, you can use the predict() method to predict a target value given a set of features. In other words, use the model, along with some unlabelled data to predict the label.

***CAUTION***: data you predict on has to be in the same shape as data you trained on.

Fortunately we already have the correct shape since we built our train and test data randomly using the Scikit Learn **train_test_split()** function. This will ensure the x_train and x_test will be the same shape. (We already verified this above), emphasis in the number of column (in this case the second in the shape function output).

In [9]:
# just to verify
x_test

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
263   63    0   0       108   269    0        1      169      1      1.8   
34    51    1   3       125   213    0        0      125      1      1.4   
96    62    0   0       140   394    0        0      157      0      1.2   
161   55    0   1       132   342    0        1      166      0      1.2   
19    69    0   3       140   239    0        1      151      0      1.8   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
115   37    0   2       120   215    0        1      170      0      0.0   
297   59    1   0       164   176    1        0       90      0      1.0   
12    49    1   1       130   266    0        1      171      0      0.6   
36    54    0   2       135   304    1        1      170      0      0.0   
121   59    1   0       138   271    0        0      182      0      0.0   

     slope  ca  thal  
263      1   2     2  
34       2   1     2  
96       1   0     2  
161      2   0     2  
19       2   2     2  
..     ...  ..   ...  
115      2   0     2  
297      1   2     1  
12       2   0     2  
36       2   0     2  
121      2   0     2  

[76 rows x 13 columns]

In [10]:
# now using the predict method to make prediction onf the test data 
y_preds = clf.predict(x_test)

#### 4. Evaluate the model
Now we've made some predictions (y_preds). we can start to use some more Scikit-Learn methods to figure out how good our models is. 

Each model or estimator has built-in score method. This method compares how well the model was able to learn the patterns between the features and labels. In other words, it returns how accurate your model is.

Using Random Forest Classifier score method. 

**score(X, y[, sample_weight])** https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.score

Return the mean accuracy on the given test data and labels.

In [11]:
clf.score(x_train, y_train)

1.0

In [12]:
clf.score(x_test, y_test)

0.8947368421052632

***NOTE:*** this numbers are the mean accuracy between two data set. After using the predictin model we can use the model on the test data sets. 

#### use the additional evaluation methods and also libs for reporting
This is part of the metrics libs of the sklearn. 

See the [Metrics and scoring: quantifying the quality of predictions](https://scikit-learn.org/stable/modules/model_evaluation.html#model-evaluation) section and the Pairwise metrics, Affinities and Kernels section of the user guide for further details.

The [sklearn.metrics module](https://scikit-learn.org/stable/modules/classes.html#module-sklearn.metrics) includes score functions, performance metrics and pairwise metrics and distance computations.

Here is the list I need to use from this lib:
1. [classification_report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report)
1. [confusion_matrix](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html#sklearn.metrics.confusion_matrix)
1. [accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score)


[3.3.2.7. Classification report](https://scikit-learn.org/stable/modules/model_evaluation.html#classification-report)

The classification_report function builds a text report showing the main classification metrics. Here is a small example with custom target_names and inferred labels:

**sklearn.metrics.classification_report(y_true, y_pred, *, labels=None, target_names=None, sample_weight=None, digits=2, output_dict=False, zero_division='warn')**[source](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html#sklearn.metrics.classification_report)

In [13]:
# import the sklearn.metric additiona evaluation and reporting methods
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# now print the classification report y_true is y_test, y_pred is well y_preds
print(classification_report(y_test, y_preds))

              precision    recall  f1-score   support

           0       0.86      0.91      0.88        33
           1       0.93      0.88      0.90        43

    accuracy                           0.89        76
   macro avg       0.89      0.90      0.89        76
weighted avg       0.90      0.89      0.90        76



From this lib I need [confusion matrix (here is the user guide on it)](https://scikit-learn.org/stable/modules/model_evaluation.html#confusion-matrix). Also here is [wikipedia page on cofusion matrix](https://en.wikipedia.org/wiki/Confusion_matrix).

**sklearn.metrics.confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None, normalize=None)**

returns: 
    C: ndarray of shape (n_classes, n_classes)

Confusion matrix whose i-th row and j-th column entry indicates the number of samples with true label being i-th class and predicted label being j-th class.

In [14]:
# use the confusion matrix with y_true = y_test and y_pred = y_preds
confusion_matrix(y_test, y_preds)

array([[30,  3],
       [ 5, 38]], dtype=int64)

Next is the [accuracy_score](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.accuracy_score.html#sklearn.metrics.accuracy_score)

**sklearn.metrics.accuracy_score(y_true, y_pred, *, normalize=True, sample_weight=None)**

Returns: score: float

If normalize == True, return the fraction of correctly classified samples (float), else returns the number of correctly classified samples (int).

The best performance is 1 with normalize == True and the number of samples with normalize == False.

*see the example in the source guide*

In [15]:
# use the accuracy_score with y_true = y_test and y_pred = y_preds
accuracy_score(y_test, y_preds)

0.8947368421052632

#### 5. Experiment to improve
The first model we build is often referred  to as baseline.

Once we have got a baseline model, like we have here, it is important to remember, this is often not the final model we will use. 

The next step in the workflow is to try to improve upon your baseline model.

To do this there are two ways to look at it: 
1. Form a model perspective 
1. From the data perspective

From model perspective maybe would involves things such as using a more complex model or tuning current model's hyper-parameters.

From data perspective, this may involves collecting more data or improving the data quality. This will give better chance of the model to learn the pattern within.

If you are already working on an existing dataset, it is often easier to try a series of model perspective experiments first then turn to data perspective if you are not getting the result you are looking for. This is sounds like our current case where we alreasy have heart disease data frame.

One thing you should be aware of is if you are tuning a model's hyper-parameters in a series of experiments, your results should always be **cross-validated**. 

Cross validation is a way of making sure that results you are getting are consistent across your training and test dataset. This must validate the use of multiple versions of training and test sets rather than just pure luck because of the order the training and test sets were created.
- try different hyper-parameters
- All different parameters should be cross validated
    - **CAUTION:** beware of cross validation for time series problems

Different models you use will have different hyper-parameters you can tune. For the case of our model the Random Forrest Classifier, we will start trying different values for n_estimators.